### PR TITLE
Add sudo functionality to the API

### PR DESCRIPTION
**Pull request by zeripath**
Adds sudo functionality similar to https://docs.gitlab.com/ee/api/#sudo

### CODE REVIEW

**Comment by lunny** [comment]
CI failed

**Comment by zeripath** [comment]
Have added a basic integration test - which after a small amount of hacking now works.

**Comment by tboerger** [comment]
I personally don't like the naming sudo. I know such a feature as impersonate

**Comment by zeripath** [comment]
I just took the name from the gitlab API. So used it to keep consistent
with that.

I've previously used "become user", "become", "as user", or "as".
"Impersonate" is also good.

Which would people prefer?


-- 
Andrew Thornton
art27@cantab.net

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
I'm ok with `sudo` :)

**Comment by zeripath** [comment]
I've just rebased to master, added some documentation and added the Sudo option to swagger as an authentication type.

Adding Sudo as an authentication type appears the easiest way to add it to the API and I guess it is a kind of authentication. Although I tried I couldn't seem to get swagger-generate to generate the file in such a way that Sudo requires another form of authentication.

File: docs/content/doc/advanced/api-usage.en-us.md [hunk 1]
```diff
@@ -73,3 +73,7 @@ using BasicAuth, as follows:
 $ curl --request GET --url https://yourusername:yourpassword@gitea.your.host/api/v1/users/yourusername/tokens
 [{"name":"test","sha1":"..."},{"name":"dev","sha1":"..."}]
 ```
+
+## Sudo
+
+The api allows admin users to sudo api requests as another user. Simply add either a `sudo=` parameter or `Sudo:` request header with the username of the user to sudo.
```
Comment (lafriks) [inline docs/content/doc/advanced/api-usage.en-us.md]: Please use `API ` in upper case in all places (docs, swagger docs etc)

Comment (zeripath) [reply docs/content/doc/advanced/api-usage.en-us.md]: Fixed.

**Review by lafriks — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Pull request merged**